File: web/src/types/device.ts
export type CapabilityType =
  | 'devices.capabilities.on_off'
  | 'devices.capabilities.color_setting'
  | 'devices.capabilities.mode'
  | 'devices.capabilities.range'
  | 'devices.capabilities.toggle'

export type PropertyType =
  | 'devices.properties.float'
  | 'devices.properties.event'

export interface CapabilityState {
  instance?: string
  value?: unknown
  action_result?: { status?: string; error_code?: string; error_message?: string }
}

export interface Capability {
  id: string
  type: CapabilityType
  retrievable: boolean
  reportable: boolean
  parameters?: Record<string, unknown>
  state?: CapabilityState
}

export interface Property {
  id: string
  type: PropertyType
  retrievable: boolean
  reportable: boolean
  parameters?: Record<string, unknown>
  state?: CapabilityState
}

export interface Device {
  id: string
  device_uid: string
  name: string
  description: string
  room: string
  type: string
  status_info?: Record<string, unknown>
  custom_data?: Record<string, unknown>
  device_info?: Record<string, unknown>
  last_seen: string
  created_at: string
  updated_at: string
  capabilities: Capability[]
  properties: Property[]
}

export interface DeviceListResponse {
  devices: Device[]
}

export interface UpdateDeviceRequest {
  name?: string
  description?: string
  room?: string
}

export interface SetCapabilityRequest {
  instance?: string
  value: unknown
}

export const CAPABILITY_LABELS: Record<CapabilityType, string> = {
  'devices.capabilities.on_off': 'Power',
  'devices.capabilities.color_setting': 'Color',
  'devices.capabilities.mode': 'Mode',
  'devices.capabilities.range': 'Level',
  'devices.capabilities.toggle': 'Toggle',
}

export const DEVICE_TYPE_LABELS: Record<string, string> = {
  'devices.types.light': 'Light',
  'devices.types.light.strip': 'LED strip',
  'devices.types.socket': 'Socket',
  'devices.types.switch': 'Switch',
  'devices.types.thermostat': 'Thermostat',
  'devices.types.sensor': 'Sensor',
  'devices.types.media_device': 'Media',
  'devices.types.other': 'Other',
}

export function deviceTypeLabel(type: string): string {
  return DEVICE_TYPE_LABELS[type] ?? type.replace('devices.types.', '')
}

export function isOn(device: Device): boolean | undefined {
  const onOff = device.capabilities.find(c => c.type === 'devices.capabilities.on_off')
  if (!onOff?.state) return undefined
  return Boolean(onOff.state.value)
}

export function getCapability(device: Device, type: CapabilityType): Capability | undefined {
  return device.capabilities.find(c => c.type === type)
}

File: devices-api/internal/models/device.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type DeviceWithRelations struct {
	ID           uuid.UUID       `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()" json:"id"`
	DeviceUID    string          `gorm:"column:device_uid" json:"device_uid"`
	MacAddress   string          `gorm:"column:mac_address" json:"-"`
	UserID       uuid.UUID       `gorm:"column:user_id" json:"-"`
	Name         string          `json:"name"`
	Description  string          `json:"description"`
	Room         string          `json:"room"`
	Type         string          `json:"type"`
	StatusInfo   json.RawMessage `gorm:"column:status_info" json:"status_info,omitempty"`
	CustomData   json.RawMessage `gorm:"column:custom_data" json:"custom_data,omitempty"`
	DeviceInfo   json.RawMessage `gorm:"column:device_info" json:"device_info,omitempty"`
	LastSeen     time.Time       `gorm:"column:last_seen" json:"last_seen"`
	CreatedAt    time.Time       `gorm:"column:created_at" json:"created_at"`
	UpdatedAt    time.Time       `gorm:"column:updated_at" json:"updated_at"`
	DeletedAt    gorm.DeletedAt  `gorm:"index" json:"-"`
	Capabilities []Capability    `gorm:"foreignKey:DeviceID" json:"capabilities"`
	Properties   []Property      `gorm:"foreignKey:DeviceID" json:"properties"`
}

func (DeviceWithRelations) TableName() string { return "devices" }

type Capability struct {
	ID          uuid.UUID       `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()" json:"id"`
	DeviceID    uuid.UUID       `gorm:"column:device_id" json:"-"`
	Type        string          `json:"type"`
	Retrievable bool            `json:"retrievable"`
	Reportable  bool            `json:"reportable"`
	Parameters  json.RawMessage `json:"parameters,omitempty"`
	State       json.RawMessage `json:"state,omitempty"`
}

func (Capability) TableName() string { return "capabilities" }

type Property struct {
	ID          uuid.UUID       `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()" json:"id"`
	DeviceID    uuid.UUID       `gorm:"column:device_id" json:"-"`
	Type        string          `json:"type"`
	Retrievable bool            `json:"retrievable"`
	Reportable  bool            `json:"reportable"`
	Parameters  json.RawMessage `json:"parameters,omitempty"`
	State       json.RawMessage `json:"state,omitempty"`
}

func (Property) TableName() string { return "properties" }

type UpdateDeviceRequest struct {
	Name        *string `json:"name"`
	Description *string `json:"description"`
	Room        *string `json:"room"`
}

type SetCapabilityRequest struct {
	Instance string          `json:"instance"`
	Value    json.RawMessage `json:"value" binding:"required"`
}

type DeviceListResponse struct {
	Devices []DeviceWithRelations `json:"devices"`
}
